rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,461 @@
+
+Historically, in `blade-old` _(deprecated)_ we followed an enhancer components based pattern which looked like below:
+
+```tsx
+<Space padding={[2.5, 0, 5, 0]}>
+  <View>
+    <Heading>Account Blocked</Heading>
+  </View>
+</Space>
+```
+
+This suffered from some related problems:
+
+1. Need for creating nested wrapper components for everything for any adjustments - `Space`, `Size` and `View`
+1. Readability wasn't as good because actual styles would get applied to the nested `View` which also made refactors harder
+1. Slightly higher learning curve
+1. No built-ins for responsive styles
+
+### Proposed solution
+
+With layout primitives in place, the above example could be implemented like:
+
+```tsx
+// one of the ways
+<Flex flexDirection="column" alignItems="center">
+  <Button>Button 1</Button>
+  <Button sx={{ marginTop: 'spacing.2' }}>Button 2</Button>
+</Flex>
+```
+
+Thus, layout primitives can ease up building UIs with:
+
+1. Avoiding need for dedicatedly creating wrapper components for majority of usecases with ability to apply spacing directly on components with `sx`, having components such as `Stack`, `Spacer` for adding spacing between siblings or max flexibility with a `Box` component
+1. Provide built-in capabilities for easing up responsive layouts with breakpoints based props such as `{ base: 'spacing.1', m: 'spacing.2' }`
+1. No need to manually pass tokens for layout styling, the primitives will be able to derive actual value from tokens like `spacing.1`, `spacing.2`, etc.
+1. Cross platform components such as `Box`, `Flex`, `Spacer` etc. that work for both React and React Native with the same API.
+
+Indubitably, layouts have been one of the most requested features by developers using Blade _([link](https://docs.google.com/document/d/14VAfBW7vz6hYOC1eWQO-_wOcECLU9UXRHVOSPC6OxZ8/edit#heading=h.zg0vxpo7xuc8) to internal document)_.
+
+## Detailed Design
+
+### Prior art
+
+Every existing design system has tried to solve layouts bit differently. However, I've observed a few trends that seem to offer the best DX:
+
+- A general purpose `Box` component along with helpful derivatives such as `Flex`, `Grid`, `Stack`, etc.
+- Having support for theme aware styling props directly on components (such as `marginTop`, `paddingRight`, etc.)
+- Reliance on having a `sx` prop for theme aware styling on components
+- Built-in support for responsive styles on theme aware properties by either accepting an array / object format for values at different breakpoints
+- Finally, some design systems also support a `style` or `className` prop on components for overrides and layouts for maximum flexibility
+
+A brief study of some layout primitives is compiled in the [references](#references) section below.
+
+### Layout primitives and components for Blade
+
+In order to provide a good balance of flexibility and constraints within blade we should support the theme aware components and primitives listed below.
+
+#### `Box` component
```

Comment:
Do we really need a generic `Box` if we are offering `Flex`, `Grid` & `Stack`?

Is there any use-case we see where `Flex`, `Grid` & `Stack` won't be enough and a developer would need to use a `Box` instead?

Reply:
For web, `Flex` might not always be a good fit. `Box` can be helpful for creating `block`, `inline-block`, `none` layouts.

On native, everything would follow a `flex` layout but not for web, so wrapping any element with `Flex` may result in unwanted changes. For example:

```tsx
<Box marginTop="spacing.2">
  {/* here using a Flex would change the child to flex instead of block */}
  <MyCustomComponent />
</Box>
```